src/storage/src/hummock/iterator/skip_watermark.rs

```diff
@@ -252,9 +319,23 @@ impl SkipWatermarkState {
                     continue;
                 }
                 Ordering::Equal => {
+                    self.last_serde = self
+                        .compaction_catalog_agent_ref
+                        .watermark_serde(table_id.table_id());
+
                     match direction {
                         WatermarkDirection::Ascending => {
-                            match inner_key.cmp(watermark.as_ref()) {
+                            let cmp = match watermark {
+                                WatermarkSerdeType::Bytes(watermark) => {
+                                    inner_key.cmp(watermark.as_ref())
+                                }
+                                WatermarkSerdeType::Serde(_watermark) => {
+                                    // do not skip the non-pk prefix watermark when vnode is the same
+                                    return false;
```

Comment:
I am afraid this is still incorrect based on the semantic of `advance_watermark`:
```
/// Return a flag indicating whether the current key will be filtered by the current watermark.
```

If we always return false when the table, vnode are the same here, that means none of the keys can be filtered by the watermark. Please clearfully walk through the logics of `advance_watermark`, `should_delete` and `advance_key_and_watermark`. I am still concerned that the implementation of `SkipWatermarkState` and `SkipWatermarkIterator` rely on the assumption that the key ordering and watermark ordering is the same and we may still miss some changes.